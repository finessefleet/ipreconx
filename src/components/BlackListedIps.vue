<template>
  <h1>IPRECONX</h1>

  <!-- 🔍 Search box -->
  <input
    type="text"
    v-model="searchTerm"
    placeholder="Search IP address..."
    class="search-box"
  />

  <!-- 🔁 Refresh button -->
  <button @click="getBlackListedIps" class="refresh-button">
    Refresh IP List
  </button>

  <!-- 📄 IP Address List -->
  <div class="container">
    <div class="ipaddress" v-for="ip in paginatedIps" :key="ip">
      {{ ip }}
    </div>
  </div>

  <!-- 📌 Pagination Controls -->
  <div class="pagination">
    <button :disabled="currentPage === 1" @click="currentPage--">Prev</button>

    <span>Page {{ currentPage }} of {{ totalPages }}</span>

    <button :disabled="currentPage === totalPages" @click="currentPage++">Next</button>
  </div>
</template>

<script setup>
import { ref, computed } from "vue"

const blacklistedIps = ref([])
const searchTerm = ref("")
const currentPage = ref(1)
const itemsPerPage = 100

const url =
  "https://raw.githubusercontent.com/firehol/blocklist-ipsets/master/firehol_level1.netset"

async function getBlackListedIps() {
  try {
    const response = await fetch(url)
    const rawContent = await response.text()
    const regex = /\b(?:\d{1,3}\.){3}\d{1,3}\b/g
    blacklistedIps.value = rawContent.match(regex) || []
    currentPage.value = 1 // reset to page 1 on refresh
  } catch (error) {
    console.error("Error fetching IP list:", error)
  }
}

// Initial fetch
getBlackListedIps()

// Filter by search
const filteredIps = computed(() =>
  blacklistedIps.value.filter(ip =>
    ip.includes(searchTerm.value.trim())
  )
)

// Total pages based on filtered result
const totalPages = computed(() =>
  Math.ceil(filteredIps.value.length / itemsPerPage)
)

// Slice for current page
const paginatedIps = computed(() => {
  const start = (currentPage.value - 1) * itemsPerPage
  const end = start + itemsPerPage
  return filteredIps.value.slice(start, end)
})
</script>

<style>
.container {
  background-color: #fff;
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(150px, 1fr));
  gap: 10px;
}

.ipaddress {
  background-color: #f1f1f1;
  padding: 7px 15px;
  border-radius: 5px;
  font-family: 'Courier New', Courier, monospace;
  color: #333;
}

.search-box {
  margin: 10px 10px 10px 0;
  padding: 8px;
  width: 100%;
  max-width: 400px;
  font-size: 16px;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.refresh-button {
  padding: 8px 12px;
  font-size: 16px;
  margin-bottom: 20px;
  background-color: #4CAF50;
  color: white;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}

.refresh-button:hover {
  background-color: #45a049;
}

.pagination {
  margin-top: 20px;
  display: flex;
  justify-content: center;
  gap: 20px;
  align-items: center;
}

.pagination button {
  padding: 6px 12px;
  font-size: 14px;
  border: none;
  background-color: #007bff;
  color: white;
  border-radius: 4px;
  cursor: pointer;
}

.pagination button:disabled {
  background-color: #ccc;
  cursor: not-allowed;
}

.pagination span {
  font-size: 16px;
}
</style>
