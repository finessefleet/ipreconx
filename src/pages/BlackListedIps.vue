<template>
  <!-- Logo displayed at the top center of the UI -->
  <img src="@/assets/images/logo.png" class="w-50 m-auto" />

  <!-- Search and Refresh Controls -->
  <div class="flex justify-center">
    <!-- Text input for user to filter IP addresses -->
    <input
      type="text"
      v-model="searchTerm" 
      placeholder="Search IP address..."
      class="bg-white border border-gray-100 px-2 py-1 rounded"
    />

    <!-- Button to re-fetch the IP list -->
    <button
      @click="getBlackListedIps" 
      class="bg-green-700 text-white py-1 px-2 border border-green-800 mx-3 rounded"
    >
      <!-- Conditional text based on loading state -->
      <span v-if="loading">Loading</span>
      <span v-else>Refresh IP List</span>
    </button>
  </div>

  <!-- Display paginated IPs if available -->
  <div class="flex flex-wrap justify-center gap-3 my-5" v-if="paginatedIps.length">
    <div class="bg-gray-100 rounded p-2 w-[150px]" v-for="ip in paginatedIps" :key="ip" @click="getIpLocation(ip)">
      {{ ip }}
    </div>
  </div>

  <!-- Message shown when no IPs match the search input -->
  <div class="text-gray-500 text-center my-10" v-else>
    No blacklisted IP address found containing "{{ searchTerm }}"
  </div>

  <!-- Pagination controls to navigate between pages -->
  <Pagination
    :currentPage="currentPage"           
    :totalPages="totalPages"            
    @update:currentPage="currentPage = $event"
  />
</template>

<script setup>
// Import Vueâ€™s Composition API methods
import { ref, computed, watch } from "vue";
// Import Pagination component (assumed to be reusable)
import Pagination from '@/components/Pagination.vue';
import { useRouter } from "vue-router";

const blacklistedIps = ref([]);
const searchTerm = ref("");
const currentPage = ref(1);
const loading = ref(false);
const itemsPerPage = 100;

const router = useRouter();
// Remote URL where the IP blacklist is hosted
const url = "https://raw.githubusercontent.com/firehol/blocklist-ipsets/master/firehol_level1.netset";

/**
 * Fetches the blacklisted IP addresses from the given remote URL.
 * Extracts IPs using regex and updates state.
 */
async function getBlackListedIps() {
  try {
    loading.value = true; // Enable loading UI

    // Fetch the raw content from the URL
    const response = await fetch(url);
    const rawContent = await response.text();

    // Extract all IPv4 addresses using regex
    const regex = /\b(?:\d{1,3}\.){3}\d{1,3}\b/g;
    blacklistedIps.value = rawContent.match(regex) || [];

    // Reset search and pagination on refresh
    searchTerm.value = '';
    currentPage.value = 1;
  } catch (error) {
    console.error("Error fetching IP list:", error);
  } finally {
    loading.value = false;
  }
}

// Automatically fetch IP list when component is first mounted
getBlackListedIps();

/**
 * Returns a filtered list of IPs that match the current search input.
 * Case-sensitive substring match is used here.
 */
const filteredIps = computed(() =>
  blacklistedIps.value.filter(ip =>
    ip.includes(searchTerm.value.trim())
  )
);

/**
 * Calculates the total number of pages based on the number of filtered IPs.
 */
const totalPages = computed(() => {
  currentPage.value = 1;
  return Math.ceil(filteredIps.value.length / itemsPerPage)
});

/**
 * Returns only the IPs that should be displayed on the current page.
 * Handles pagination slicing logic.
 */
const paginatedIps = computed(() => {
  const start = (currentPage.value - 1) * itemsPerPage
  const end = start + itemsPerPage
  return filteredIps.value.slice(start, end)
})

const getIpLocation = (ipaddress) => {
  // https://www.virustotal.com/api/v3/ip_addresses/{ip}
  fetch(`https://api.iplocation.net/?ip=${ipaddress}`).then(async res => {
    const {country_name, isp} = await res.json();
    alert(country_name + ' : ' + isp);
  });
  router.push({name: 'ip', params: { ip: ipaddress}})
}
</script>
