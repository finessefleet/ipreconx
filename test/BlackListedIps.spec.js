import { describe, it, expect, vi, beforeEach } from 'vitest'
import { mount, flushPromises } from '@vue/test-utils'
import BlackListedIps from '/src/components/BlackListedIps.vue'


// Mock fetch
global.fetch = vi.fn()

const mockIps = `
1.1.1.1
2.2.2.2
3.3.3.3
4.4.4.4
5.5.5.5
`

describe('BlackListedIps.vue', () => {
  beforeEach(() => {
    fetch.mockReset()
  })

  it('fetches and displays blacklisted IPs', async () => {
    fetch.mockResolvedValue({
      text: () => Promise.resolve(mockIps),
    })

    const wrapper = mount(BlackListedIps)

    // Wait for async operations to finish
    await flushPromises()

    // Check IPs are rendered
    expect(wrapper.text()).toContain('1.1.1.1')
    expect(wrapper.text()).toContain('5.5.5.5')

    // Should render 5 IPs
    const ipElements = wrapper.findAll('div.bg-gray-100')
    expect(ipElements.length).toBe(5)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('filters IPs based on search term', async () => {
    fetch.mockResolvedValue({
      text: () => Promise.resolve(mockIps),
    })

    const wrapper = mount(BlackListedIps)
    await flushPromises()

    // Set search term
    await wrapper.find('input').setValue('2.2')

    // Only 2.2.2.2 should match
    expect(wrapper.text()).toContain('2.2.2.2')
    expect(wrapper.text()).not.toContain('1.1.1.1')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('shows message when no IPs match', async () => {
    fetch.mockResolvedValue({
      text: () => Promise.resolve(mockIps),
    })

    const wrapper = mount(BlackListedIps)
    await flushPromises()

    await wrapper.find('input').setValue('9.9.9.9')

    expect(wrapper.text()).toContain('No blacklisted IP address found containing 9.9.9.9')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('resets loading state after fetch', async () => {
    vi.useFakeTimers()
    fetch.mockResolvedValue({
      text: () => Promise.resolve(mockIps),
    })

    const wrapper = mount(BlackListedIps)
    await flushPromises()

    // Simulate clicking the refresh button
    await wrapper.find('button').trigger('click')

    // Fast forward timers
    vi.runAllTimers()

    // loading should be false
    expect(wrapper.text()).not.toContain('Loading')

    vi.useRealTimers()
  })
})
