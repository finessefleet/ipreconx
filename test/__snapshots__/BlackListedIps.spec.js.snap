// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BlackListedIps.vue > fetches and displays blacklisted IPs 1`] = `
"<img src="/src/assets/images/logo.png" class="w-50 m-auto">
<div class="flex justify-center"><input type="text" placeholder="Search IP address..." class="bg-white border border-gray-100 px-2 py-1 rounded"><button class="bg-green-700 text-white py-1 px-2 border border-green-800 mx-3 rounded"><span>Refresh IP List</span></button></div>
<div class="flex flex-wrap justify-center gap-3 my-5">
  <div class="bg-gray-100 rounded p-2 w-[150px]">1.1.1.1</div>
  <div class="bg-gray-100 rounded p-2 w-[150px]">2.2.2.2</div>
  <div class="bg-gray-100 rounded p-2 w-[150px]">3.3.3.3</div>
  <div class="bg-gray-100 rounded p-2 w-[150px]">4.4.4.4</div>
  <div class="bg-gray-100 rounded p-2 w-[150px]">5.5.5.5</div>
</div>
<!--v-if-->"
`;

exports[`BlackListedIps.vue > filters IPs based on search term 1`] = `
"<img src="/src/assets/images/logo.png" class="w-50 m-auto">
<div class="flex justify-center"><input type="text" placeholder="Search IP address..." class="bg-white border border-gray-100 px-2 py-1 rounded"><button class="bg-green-700 text-white py-1 px-2 border border-green-800 mx-3 rounded"><span>Refresh IP List</span></button></div>
<div class="flex flex-wrap justify-center gap-3 my-5">
  <div class="bg-gray-100 rounded p-2 w-[150px]">2.2.2.2</div>
</div>
<!--v-if-->"
`;

exports[`BlackListedIps.vue > shows message when no IPs match 1`] = `
"<img src="/src/assets/images/logo.png" class="w-50 m-auto">
<div class="flex justify-center"><input type="text" placeholder="Search IP address..." class="bg-white border border-gray-100 px-2 py-1 rounded"><button class="bg-green-700 text-white py-1 px-2 border border-green-800 mx-3 rounded"><span>Refresh IP List</span></button></div>
<div class="text-gray-100 text-center my-50"> No blacklisted IP address found containing 9.9.9.9</div>
<!--v-if-->"
`;
